qadence/qinfo_tools/qfi.py

```diff
@@ -0,0 +1,261 @@
+    the fidelity of the quantum state
+
+    Args:
+        circuit (QuantumCircuit): _description_
+        backend (BackendName, optional): _description_. Defaults to BackendName.pyq.
+        overlap_method (OverlapMethod, optional): _description_. Defaults to OverlapMethod.EXACT.
+        diff_mode (DiffMode, optional): _description_. Defaults to DiffMode.ad.
+
+    Returns:
+        torch.Tensor: QFI matrix
+    """
+    num_vparams = len(var_values) if var_values is not None else 0
+
+    # Random directions
+    Delta_1 = torch.Tensor(np.random.choice([-1, 1], size=(num_vparams, 1)))
+    Delta_2 = torch.Tensor(np.random.choice([-1, 1], size=(num_vparams, 1)))
+
+    ovrlp_model = Overlap(
+        circuit,
+        circuit,
+        backend=backend,
+        diff_mode=diff_mode,
+        method=overlap_method,
+    )
+
+    if fm_dict == {}:
+        blocks = primitive_blocks(circuit.block)
+        for block in blocks:
+            params = parameters(block)
+            for p in params:
+                if not p.trainable:
+                    fm_dict[p.name] = torch.Tensor([p.value])
+
+    # Calculate the shifted parameters
+    vparams_tensors = torch.Tensor(var_values).reshape((num_vparams, 1))
+    vparams = {k: v for (k, v) in ovrlp_model._params.items() if v.requires_grad}
+    vparams_plus1_plus2 = dict(zip(vparams.keys(), vparams_tensors + epsilon * (Delta_1 + Delta_2)))
+    vparams_plus1 = dict(zip(vparams.keys(), vparams_tensors + epsilon * (Delta_1)))
+    vparams_minus1_plus2 = dict(
+        zip(vparams.keys(), vparams_tensors + epsilon * (-Delta_1 + Delta_2))
+    )
+    vparams_minus1 = dict(zip(vparams.keys(), vparams_tensors + epsilon * (-Delta_1)))
+
+    # Overlaps with the shifted parameters
+    ovrlp_shifter_plus1_plus2 = ovrlp_model(
+        bra_param_values=fm_dict | vparams,
+        ket_param_values=fm_dict | vparams_plus1_plus2,
+    )
+    ovrlp_shifter_plus1 = ovrlp_model(
+        bra_param_values=fm_dict | vparams, ket_param_values=fm_dict | vparams_plus1
+    )
+    ovrlp_shifter_minus1_plus2 = ovrlp_model(
+        bra_param_values=fm_dict | vparams,
+        ket_param_values=fm_dict | vparams_minus1_plus2,
+    )
+    ovrlp_shifter_minus1 = ovrlp_model(
+        bra_param_values=fm_dict | vparams, ket_param_values=fm_dict | vparams_minus1
+    )
+
+    delta_F = (
+        ovrlp_shifter_plus1_plus2
+        - ovrlp_shifter_plus1
+        - ovrlp_shifter_minus1_plus2
+        + ovrlp_shifter_minus1
+    )
+
+    fid_hess = (
+        (1 / 4)
+        * (delta_F / (epsilon**2))
+        * (
+            torch.matmul(Delta_1, Delta_2.transpose(0, 1))
+            + torch.matmul(Delta_2, Delta_1.transpose(0, 1))
+        )
+    )
+    qfi_mat = -2 * fid_hess
+
+    # Calculate the new estimator from the old estimator of qfi_mat
+    if k == 0:
+        qfi_mat_estimator = qfi_mat
+    else:
+        qfi_mat_estimator = (1 / (k + 1)) * (k * previous_qfi_estimator + qfi_mat)  # type: ignore
```

Comment:
This previous_qfi_estimator isn't defined anywhere it seems.

Reply:
True! I think the docstrings need to be updated in general, I will check them all and make sure they are updated.

`previous_qfi_estimator` is an input to the `get_quantum_fisher_spsa()`. This function computes the estimator of the QFI matrix corresponding to the iteration k (also an input to the function). You need the estimator from the previous iteration (k-1) to construct the new estimator.